File: java-pine/src/main/java/com/tipsy/pine/operators/TransformNormalize.java
package com.tipsy.pine.operators;

import com.tipsy.pine.AbstractOperator;
import com.tipsy.pine.OperatorInput;
import com.tipsy.pine.OperatorOutput;

import java.util.Map;

public class TransformNormalize extends AbstractOperator {
    private String method;

    @Override
    public void init(Map<String, Object> params) throws Exception {
        method = (String) params.getOrDefault("method", "min_max");
        if (!"min_max".equals(method)) {
            throw new IllegalArgumentException("transform_normalize: unsupported method: " + method);
        }
    }

    @Override
    public void execute(OperatorInput input, OperatorOutput output) throws Exception {
        int n = input.itemCount();
        if (n == 0) return;

        String field = itemInput.get(0);
        String outputField = itemOutput.get(0);

        double[] vals = new double[n];
        for (int i = 0; i < n; i++) {
            vals[i] = toDouble(input.item(i, field));
        }

        double min = vals[0], max = vals[0];
        for (int i = 1; i < n; i++) {
            if (vals[i] < min) min = vals[i];
            if (vals[i] > max) max = vals[i];
        }

        double range = max - min;
        for (int i = 0; i < n; i++) {
            double norm = range == 0 ? 0.0 : (vals[i] - min) / range;
            output.setItem(i, outputField, norm);
        }
    }

    private static double toDouble(Object v) throws Exception {
        if (v instanceof Number) {
            return ((Number) v).doubleValue();
        }
        throw new Exception("cannot convert " + (v == null ? "null" : v.getClass().getName()) + " to double");
    }
}

File: pine-cpp/operators/transform/transform_normalize.cpp
#include "pine/operator.hpp"

#include <algorithm>

#include "operators/_helpers.hpp"

namespace pine {

class TransformNormalizeOp : public Operator {
 public:
  void init(const OperatorConfig& cfg) override {
    op_name_ = cfg.name;
    item_input_ = cfg.metadata.item_input;
    item_output_ = cfg.metadata.item_output;
    std::string method = "min_max";
    const auto& params = cfg.params.as_object();
    auto it = params.find("method");
    if (it != params.end() && !it->second.is_null()) {
      method = it->second.as_string();
    }
    if (method != "min_max") {
      throw RegistryError(op_name_, "unsupported method \"" + method + "\"");
    }
  }
  void execute(const OperatorInput& input, OperatorOutput& out) override {
    if (input.item_count() == 0) {
      return;
    }
    if (item_input_.empty()) {
      throw ExecutionError(op_name_, "transform_normalize: item_input is empty");
    }
    if (item_output_.empty()) {
      throw ExecutionError(op_name_, "transform_normalize: item_output is empty");
    }
    const std::string& field_ = item_input_[0];
    const std::string& out_field_ = item_output_[0];
    // Batched column access: one lock + one lookup instead of per-element
    // item() calls.
    std::vector<Variant> raw = input.item_column(field_);
    std::vector<double> vals;
    vals.reserve(input.item_count());
    for (std::size_t i = 0; i < raw.size(); ++i) {
      try {
        const Variant& v = raw[i];
        if (v.is_null()) {
          throw operators::OperatorError("required field \"" + field_ + "\" is nil on item[" +
                                         std::to_string(i) + "]");
        }
        vals.push_back(operators::to_double(v));
      } catch (const operators::OperatorError& err) {
        throw ExecutionError("transform_normalize: item[" + std::to_string(i) + "]." + field_ + ": " +
                             err.what());
      }
    }
    double minv = *std::min_element(vals.begin(), vals.end());
    double maxv = *std::max_element(vals.begin(), vals.end());
    double rng = maxv - minv;
    for (std::size_t i = 0; i < vals.size(); ++i) {
      double norm = (rng == 0.0) ? 0.0 : (vals[i] - minv) / rng;
      out.set_item(static_cast<int>(i), out_field_, Variant(norm));
    }
  }

 private:
  std::string op_name_;
  std::vector<std::string> item_input_;
  std::vector<std::string> item_output_;
};

static const OperatorSchema k_transform_normalize_schema{
    .name = "transform_normalize",
    .type = OpType::Transform,
    .description = "Normalizes a numeric item field using min-max scaling to [0, 1].",
    .params =
        {
            {"method",
             {.type = "string",
              .required = false,
              .default_value = Variant("min_max"),
              .description = "Normalization method."}},
        },
    .metadata = {.common_input = "[]",
                 .common_output = "[]",
                 .item_input = "[<field>]",
                 .item_output = "[<output_field>]"},
};
PINE_REGISTER_OPERATOR_T(TransformNormalizeOp, k_transform_normalize_schema)

}  // namespace pine
